src/TraceEvent/Computers/TraceProcess.cs

```diff
@@ -342,21 +350,79 @@ internal TraceProcess GetOrCreateProcess(int processID, long timeQPC, bool isPro
+                    // a lifetime policy is being applied which means that we may have removed
+                    //  processes leading to holes
+                    // find an available hole - linear search (list should be short)
+                    for(processIndex = 0; processIndex<processes.Count; processIndex++)
+                    {
+                        if (processes[processIndex] == null) break;
+                    }
+                }
+                retProcess = new TraceProcess(processID, log, (ProcessIndex)processIndex, source);
                 retProcess.firstEventSeenQPC = timeQPC;
-                processes.Add(retProcess);
+                if (processIndex < processes.Count) processes[processIndex] = retProcess;
+                else processes.Add(retProcess);
                 processesByPID.Insert(index + 1, retProcess);
                 // fire event
                 FireOnInitialized(retProcess);
             }
             return retProcess;
         }
+        internal void TerminateProcess(ProcessTraceData data)
+        {
+            int index;
+            FindProcessAndIndex(data.ProcessID, data.TimeStampQPC, out index);
+            var process = processesByPID[index];
+
+            // end this process
+            process.ProcessEnd(data);
+
+            // check if we should apply a lifetime policy
+            if (source.DataLifetimeMsec > 0)
+            {
+                // check processes if they are outside of the requested lifetime and remove them
+                // not relying on the process/end event since it is not always the last event
+                index = 0;
+                while (index < processesByPID.Count)
+                {
+                    process = processesByPID[index];
+
+                    System.Diagnostics.Debug.Assert(process != null);
+
+                    // if a process has had an event within the last LifetimeMsec, then keep it
+                    if (process.EndTimeRelativeMsec >= (data.TimeStampRelativeMSec - source.DataLifetimeMsec))
+                    {
+                        index++;
+                    }
+                    else
+                    {
+                        // fire event
+                        FireOnRemoved(process);
+
+                        // remove from processesByPID - this is a shift copy operation, ugh
+                        processesByPID.RemoveRange(index, 1);
+                        // remove this process from the stream and return this processIndex back to rotation
+                        processes[(int)process.ProcessIndex] = null;
+
+                        // do not advance index, since we should inspect the one that was shifted into this spot
+                    }
+                }
+            }
+        }
 
         internal void FireOnInitialized(TraceProcess proc)
         {
             if (OnInitialized != null) OnInitialized(proc);
         }
 
+        internal void FireOnRemoved(TraceProcess proc)
```

Comment:
nit, I would not bother with this method  and just inline the body (frankly I would get rid of FireOnInitlialized()) for the same reason.

Reply:
Done.  I also removed FireOnInitialized to be consistent.